components/brave_wallet/browser/internal/hd_key_ed25519_slip23.cc

```diff
@@ -0,0 +1,237 @@
+/* Copyright (c) 2025 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/internal/hd_key_ed25519_slip23.h"
+
+#include <array>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "base/check.h"
+#include "base/containers/span.h"
+#include "base/containers/span_writer.h"
+#include "base/memory/ptr_util.h"
+#include "base/strings/string_number_conversions.h"
+#include "crypto/hmac.h"
+#include "third_party/boringssl/src/include/openssl/curve25519.h"
+#include "third_party/boringssl/src/include/openssl/evp.h"
+
+namespace brave_wallet {
+
+namespace {
+
+// https://datatracker.ietf.org/doc/html/rfc8032#section-5.1.5
+// requires scalar to follow this requirements 'The lowest 3 bits of the first
+// octet are cleared, the highest bit of the last octet is cleared, and the
+// second highest bit of the last octet is set'.
+// https://input-output-hk.github.io/adrestia/static/Ed25519_BIP.pdf extends
+// this requirement to `We admit only those k such that the third highest bit
+// of the last byte of k is zero`
+base::span<uint8_t, kSlip23ScalarSize> ClampScalarEd25519Bip32(
+    base::span<uint8_t, kSlip23ScalarSize> scalar) {
+  scalar[0] &= 0b1111'1000;  // The lowest 3 bits of the first
+                             // octet are cleared
+
+  scalar[31] &= 0b0111'1111;  // highest bit of the last octet is cleared and
+  scalar[31] &= 0b1101'1111;  // third highest bit of the last byte of k is zero
+
+  scalar[31] |= 0b0100'0000;  // the second highest bit of the last octet is set
+  return scalar;
+}
+
+// https://datatracker.ietf.org/doc/html/rfc8032#section-5.1.5
+// requires scalar to follow this requirements 'The lowest 3 bits of the first
+// octet are cleared, the highest bit of the last octet is cleared, and the
+// second highest bit of the last octet is set'.
+bool IsValidEd25519Scalar(base::span<const uint8_t, kSlip23ScalarSize> scalar) {
```

Comment:
Is there a way to reuse IsScalarPruned here rather than reimplement the logic? I see these are effectively the same function so I wasn't sure what the reasoning for repeating it was here.

Reply:
removed copypaste